### PR TITLE
Build system reimplementation

### PR DESCRIPTION
**Pull request by czarkoff**
This is a followup to PR #156. This commit introduces new build system, completely from scratch:

* `Makefile` in distribution's base directory identifies operating system type (one of "osx", "unix" and "windows"), creates build directory "build-*os*-*arch*", `cd`s there and calls appropriate makefile. Operating system type may be provided explicitly with `OSTYPE` make flag.
* `mk/osx.mk`, `mk/unix.mk` and `mk/windows.mk` provide build systems for OS X, traditional Unix-like systems and Windows. All of them include `mk/common.mk`, which contains common instructions.

This build system should survive addition of other platforms if needed. It supports separate build with read-only source directory (`mkdir builddir && cd builddir && make -f /path/to/utox`). It provides no facilities for cross-compilation, but I guess it should work.

As a side bonus, the mess of OSX scripts is gone. I would prefer to squash it down further, but I don't have access to OSX at the moment. I guess it doesn't really need to build all-shared binary, but that decision is left as an excercise for reader.

[Review on Reviewable](https://reviewable.io/reviews/grayhatter/utox/178)

### CODE REVIEW

**Comment by cebe** [comment]
works for me but does not re-make on header file change like it was fixed in #77

**Comment by cebe** [comment]
I have sent a pull request to your branch which fixes this: https://github.com/czarkoff/uTox/pull/1

**Comment by czarkoff** [comment]
@cebe: I did not include that on purpose. This check that you add does no good - it invalidates all objects in case of any header change. There are two ways to do it right:
* set and maintain explicit dependency lines between all sources and headers or
* type `make clean build` instead of `make`.

Invalidating windows objects after change in xlib headers is just wrong.

**Comment by GrayHatter** [comment]
@czarkoff it's actually the right way for µTox. There are some program wide macros in other headers. Including but not limited to the Langs files, that would need everything to be recompiled when changed.

Unless you're willing to go through every file and set the dependencies by hand. You should pull @cebe's change.

**Comment by czarkoff** [comment]
@GrayHatter @cebe: Well, I see why you do that. Sad.

**Comment by GrayHatter** [comment]
@czarkoff it's not detecting windows cygwin correctly

File: mk/unix.mk [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ifeq (${OS},Linux)
+OS_CFLAGS +=	-DLINUX_IO
+OS_LDFLAGS +=	-ldl
+endif
+
+ifneq (${OS},OpenBSD)
+OS_LDFLAGS +=	-lresolv
+endif
+
+DIRs :=		png xlib
```
Comment (benwaffle) [inline mk/unix.mk]: thats a lot of room

**Comment by benwaffle** [comment]
you should use spaces, not tabs

**Comment by GrayHatter** [comment]
can't Makefile requires tabs

**Comment by benwaffle** [comment]
i mean for indenting stuff like unix.mk:8

**Comment by czarkoff** [comment]
@benwaffle tabs require less whitespace change.

@GrayHatter More details?

**Comment by GrayHatter** [comment]
@czarkoff working on a patch now

**Comment by GrayHatter** [comment]
@czarkoff install needs to be moved into unix.mk as there's no reason to have that in common

**Comment by benwaffle** [comment]
>tabs require less whitespace change.

what does that mean?

if you look in the files changed tab for your PR you will see that it looks messed up. you can only use spaces for aligning things, because people set up tabs to different widths

**Comment by czarkoff** [comment]
@benwaffle If people want to screw up their tabs, who am I to stand in their way? If they don't, they should never change tab width. Spaces are not for shifting.

**Comment by GrayHatter** [comment]
> @benwaffle If people want to screw up their tabs, who am I to stand in their way? If they don't, they should never change tab width. Spaces are not for shifting.

tab = 4 spaces, anyone who says different is wrong.

Also, tabs are better than spaces only when you're short on byte space... That's not the case anymore.

**Comment by benwaffle** [comment]
> If people want to screw up their tabs

github uses 8-space tabs. some people like 2 or 3. 
all i'm saying is use spaces to align things, not for indentation

**Comment by benwaffle** [comment]
do you see the problem
screenshot from 2016-01-20 16-24-08

**Comment by GrayHatter** [comment]
@czarkoff realpath doesn't exist on windows... weren't you complaining about using system specific commands and programs?

**Comment by czarkoff** [comment]
@benwaffle This is because of leading "+".

@GrayHatter It doesn't exist elsewhere as well. It is a GNU make macro.

**Comment by GrayHatter** [comment]
@czarkoff then SRCDIR is getting clobbered somewhere

**Comment by czarkoff** [comment]
@GrayHatter What does `make info` show you?

**Comment by GrayHatter** [comment]
@czarkoff 

```
make[1]: Entering directory '/cygdrive/c/Users/grayhatter/code/utox'
Package dbus-1 was not found in the pkg-config search path.
Perhaps you should add the directory containing `dbus-1.pc'
to the PKG_CONFIG_PATH environment variable
No package 'dbus-1' found
Package libv4lconvert was not found in the pkg-config search path.
Perhaps you should add the directory containing `libv4lconvert.pc'
to the PKG_CONFIG_PATH environment variable
No package 'libv4lconvert' found
Package dbus-1 was not found in the pkg-config search path.
Perhaps you should add the directory containing `dbus-1.pc'
to the PKG_CONFIG_PATH environment variable
No package 'dbus-1' found
Package libv4lconvert was not found in the pkg-config search path.
Perhaps you should add the directory containing `libv4lconvert.pc'
to the PKG_CONFIG_PATH environment variable
No package 'libv4lconvert' found
Package dbus-1 was not found in the pkg-config search path.
Perhaps you should add the directory containing `dbus-1.pc'
to the PKG_CONFIG_PATH environment variable
No package 'dbus-1' found
Package libv4lconvert was not found in the pkg-config search path.
Perhaps you should add the directory containing `libv4lconvert.pc'
to the PKG_CONFIG_PATH environment variable
No package 'libv4lconvert' found
Package dbus-1 was not found in the pkg-config search path.
Perhaps you should add the directory containing `dbus-1.pc'
to the PKG_CONFIG_PATH environment variable
No package 'dbus-1' found
Package libv4lconvert was not found in the pkg-config search path.
Perhaps you should add the directory containing `libv4lconvert.pc'
to the PKG_CONFIG_PATH environment variable
No package 'libv4lconvert' found
 CC            x86_64-w64-mingw32-gcc
 CFLAGS         -pthread -std=gnu99 -DGIT_VERSION="v0.5.1-10-g6b13d927"
 LDFLAGS
 Dependencies  openal libtoxcore libtoxav vpx libsodium dbus-1 libv4lconvert
 OSTYPE        windows
 SRCDIR        /cygdrive/c/Users/grayhatter/code/utox
 SRCs          /cygdrive/c/Users/grayhatter/code/utox/src/video.c /cygdrive/c/Users/grayhatter/code/utox/src/file_transfers.c /cygdrive/c/Users/grayhatter/code/utox/src/text.c /cygdrive/c/Users/grayhatter/code/utox/src/roster.c /cygdrive/c/Users/grayhatter/code/utox/src/contextmenu.c /cygdrive/c/Users/grayhatter/code/utox/src/tox.c /cygdrive/c/Users/grayhatter/code/utox/src/ui_dropdown.c /cygdrive/c/Users/grayhatter/code/utox/src/messages.c /cygdrive/c/Users/grayhatter/code/utox/src/button.c /cygdrive/c/Users/grayhatter/code/utox/src/ui_i18n.c /cygdrive/c/Users/grayhatter/code/utox/src/commands.c /cygdrive/c/Users/grayhatter/code/utox/src/edit.c /cygdrive/c/Users/grayhatter/code/utox/src/util.c /cygdrive/c/Users/grayhatter/code/utox/src/audio.c /cygdrive/c/Users/grayhatter/code/utox/src/groups.c /cygdrive/c/Users/grayhatter/code/utox/src/dns.c /cygdrive/c/Users/grayhatter/code/utox/src/ui.c /cygdrive/c/Users/grayhatter/code/utox/src/dropdown.c /cygdrive/c/Users/grayhatter/code/utox/src/scrollable.c /cygdrive/c/Users/grayhatter/code/utox/src/utox_av.c /cygdrive/c/Users/grayhatter/code/utox/src/main.c /cygdrive/c/Users/grayhatter/code/utox/src/svg.c /cygdrive/c/Users/grayhatter/code/utox/src/friend.c /cygdrive/c/Users/grayhatter/code/utox/src/theme.c /cygdrive/c/Users/grayhatter/code/utox/src/avatar.c /cygdrive/c/Users/grayhatter/code/utox/src/tooltip.c /cygdrive/c/Users/grayhatter/code/utox/src/png/png.c /cygdrive/c/Users/grayhatter/code/utox/src/windows/main.XP.c /cygdrive/c/Users/grayhatter/code/utox/src/windows/main.7.c /cygdrive/c/Users/grayhatter/code/utox/src/windows/dnd.c /cygdrive/c/Users/grayhatter/code/utox/src/windows/audio.c /cygdrive/c/Users/grayhatter/code/utox/src/windows/main.c
 HEADERS       /cygdrive/c/Users/grayhatter/code/utox/src/friend.h /cygdrive/c/Users/grayhatter/code/utox/src/theme.h /cygdrive/c/Users/grayhatter/code/utox/src/utox_av.h /cygdrive/c/Users/grayhatter/code/utox/src/avatar.h /cygdrive/c/Users/grayhatter/code/utox/src/tooltip.h /cygdrive/c/Users/grayhatter/code/utox/src/dns.h /cygdrive/c/Users/grayhatter/code/utox/src/file_transfers.h /cygdrive/c/Users/grayhatter/code/utox/src/roster.h /cygdrive/c/Users/grayhatter/code/utox/src/contextmenu.h /cygdrive/c/Users/grayhatter/code/utox/src/ui.h /cygdrive/c/Users/grayhatter/code/utox/src/ui_edits.h /cygdrive/c/Users/grayhatter/code/utox/src/ui_i18n_decls.h /cygdrive/c/Users/grayhatter/code/utox/src/main.h /cygdrive/c/Users/grayhatter/code/utox/src/messages.h /cygdrive/c/Users/grayhatter/code/utox/src/button.h /cygdrive/c/Users/grayhatter/code/utox/src/tox_bootstrap.h /cygdrive/c/Users/grayhatter/code/utox/src/theme_tables.h /cygdrive/c/Users/grayhatter/code/utox/src/ui_i18n.h /cygdrive/c/Users/grayhatter/code/utox/src/commands.h /cygdrive/c/Users/grayhatter/code/utox/src/edit.h /cygdrive/c/Users/grayhatter/code/utox/src/unused.h /cygdrive/c/Users/grayhatter/code/utox/src/svg.h /cygdrive/c/Users/grayhatter/code/utox/src/ui_dropdown.h /cygdrive/c/Users/grayhatter/code/utox/src/audio.h /cygdrive/c/Users/grayhatter/code/utox/src/groups.h /cygdrive/c/Users/grayhatter/code/utox/src/ui_buttons.h /cygdrive/c/Users/grayhatter/code/utox/src/text.h /cygdrive/c/Users/grayhatter/code/utox/src/dropdown.h /cygdrive/c/Users/grayhatter/code/utox/src/scrollable.h /cygdrive/c/Users/grayhatter/code/utox/src/tox.h /cygdrive/c/Users/grayhatter/code/utox/src/sized_string.h /cygdrive/c/Users/grayhatter/code/utox/src/util.h /cygdrive/c/Users/grayhatter/code/utox/src/video.h /cygdrive/c/Users/grayhatter/code/utox/src/tox_callbacks.h /cygdrive/c/Users/grayhatter/code/utox/src/android/main.h /cygdrive/c/Users/grayhatter/code/utox/src/cocoa/main.h /cygdrive/c/Users/grayhatter/code/utox/src/cocoa/objc_main.h /cygdrive/c/Users/grayhatter/code/utox/src/cocoa/cursor.h /cygdrive/c/Users/grayhatter/code/utox/src/png/png.h /cygdrive/c/Users/grayhatter/code/utox/src/windows/main.h /cygdrive/c/Users/grayhatter/code/utox/src/xlib/freetype.h /cygdrive/c/Users/grayhatter/code/utox/src/xlib/main.h /cygdrive/c/Users/grayhatter/code/utox/src/xlib/keysym2ucs.h /cygdrive/c/Users/grayhatter/code/utox/src/xlib/mmenu.h /cygdrive/c/Users/grayhatter/code/utox/langs/cs.h /cygdrive/c/Users/grayhatter/code/utox/langs/hi.h /cygdrive/c/Users/grayhatter/code/utox/langs/es.h /cygdrive/c/Users/grayhatter/code/utox/langs/de.h /cygdrive/c/Users/grayhatter/code/utox/langs/hu.h /cygdrive/c/Users/grayhatter/code/utox/langs/tw.h /cygdrive/c/Users/grayhatter/code/utox/langs/ru.h /cygdrive/c/Users/grayhatter/code/utox/langs/pl.h /cygdrive/c/Users/grayhatter/code/utox/langs/fr.h /cygdrive/c/Users/grayhatter/code/utox/langs/da.h /cygdrive/c/Users/grayhatter/code/utox/langs/tr.h /cygdrive/c/Users/grayhatter/code/utox/langs/ja.h /cygdrive/c/Users/grayhatter/code/utox/langs/cn.h /cygdrive/c/Users/grayhatter/code/utox/langs/eo.h /cygdrive/c/Users/grayhatter/code/utox/langs/ro.h /cygdrive/c/Users/grayhatter/code/utox/langs/ua.h /cygdrive/c/Users/grayhatter/code/utox/langs/sv.h /cygdrive/c/Users/grayhatter/code/utox/langs/pt.h /cygdrive/c/Users/grayhatter/code/utox/langs/nl.h /cygdrive/c/Users/grayhatter/code/utox/langs/bg.h /cygdrive/c/Users/grayhatter/code/utox/langs/en.h /cygdrive/c/Users/grayhatter/code/utox/langs/br.h /cygdrive/c/Users/grayhatter/code/utox/langs/no.h /cygdrive/c/Users/grayhatter/code/utox/langs/it.h /cygdrive/c/Users/grayhatter/code/utox/langs/lv.h
 OBJs          utox-icon.o video.o file_transfers.o text.o roster.o contextmenu.o tox.o ui_dropdown.o messages.o button.o ui_i18n.o commands.o edit.o util.o audio.o groups.o dns.o ui.o dropdown.o scrollable.o utox_av.o main.o svg.o friend.o theme.o avatar.o tooltip.o png/png.o windows/main.XP.o windows/main.7.o windows/dnd.o windows/audio.o windows/main.o
make[1]: Leaving directory '/cygdrive/c/Users/grayhatter/code/utox'
```

**Comment by czarkoff** [comment]
That looks right. Why do you think it is clobbered?

**Comment by GrayHatter** [comment]
@czarkoff @benwaffle you're tabs are a nightmare to read!

**Comment by GrayHatter** [comment]
@czarkoff https://gist.github.com/02a5c4279e875d4e2873

**Comment by GrayHatter** [comment]
@czarkoff https://gist.github.com/c83f681fedbd60ebd5f4

**Comment by czarkoff** [comment]
Is that with f31507e? In original version LDFLAGS for static builds were wiped on everything but OSX.

**Comment by GrayHatter** [comment]
@czarkoff yeah, and I updated the gists too

**Comment by czarkoff** [comment]
@GrayHatter FWIW my pkg-config doesn't have weird prefix. Are you sure it is supposed to be called by this lengthy name?

**Comment by GrayHatter** [comment]
@czarkoff unless I goofed, then yeah. mingw64 needs it's own version of pkg-config. and it should only be in windows.mk, unless I goofed?

**Comment by czarkoff** [comment]
@GrayHatter Well, that's it for now. Unfortunately, I won't be able to work on this PR until February, so you have the choice of merging it and fixing remaining issues or leaving it rot.

P.S.: As far as I can tell, the issues on Windows come from ancient version of GNU make that is shipped there. I guess they can be worked around.

**Comment by GrayHatter** [comment]
I'll see what I can do, I might strip out some stuff and see if I can get
that to work. But it likely wont' be tonight.

**Comment by GrayHatter** [comment]
@czarkoff should I leave this open, or close it for now?

**Comment by czarkoff** [comment]
@GrayHatter I don't have time for this right now, so do whatever you want. Once I get some free time, I may get back to this.

**Comment by GrayHatter** [comment]
okay ping me and let me know

**Pull request closed without merging**